File: app/src/main/java/remix/myplayer/ui/adapter/BaseAdapter.java
package remix.myplayer.ui.adapter;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import java.lang.reflect.Constructor;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.List;
import remix.myplayer.misc.interfaces.OnItemClickListener;
import remix.myplayer.ui.adapter.holder.BaseViewHolder;

/**
 * @ClassName
 * @Description
 * @Author Xiaoborui
 * @Date 2016/10/19 11:31
 */
public abstract class BaseAdapter<D, T extends BaseViewHolder> extends RecyclerView.Adapter<T> {

  protected OnItemClickListener mOnItemClickListener;
  protected List<D> mDatas;
  private int mLayoutId;
  private Constructor mConstructor;


  public BaseAdapter(int layoutId) {
    this.mLayoutId = layoutId;
    try {
      this.mConstructor = getGenericClass().getDeclaredConstructor(View.class);
      this.mConstructor.setAccessible(true);
    } catch (Exception e) {
      throw new IllegalArgumentException(e.toString());
    }
  }

  public void setData(List<D> datas) {
    mDatas = datas;
    notifyDataSetChanged();
  }

  public List<D> getDatas() {
    return mDatas;
  }

  @Override
  public void onBindViewHolder(T holder, int position) {
    convert(holder, getItem(position), position);
  }

  protected D getItem(int position) {
    return mDatas.get(position);
  }

  protected abstract void convert(final T holder, D d, final int position);

  @SuppressWarnings("unchecked")
  @Override
  public T onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    try {
      View itemView = LayoutInflater.from(parent.getContext()).inflate(mLayoutId, parent, false);
      return (T) mConstructor.newInstance(itemView);
    } catch (Exception e) {
      throw new IllegalArgumentException(e.toString());
    }
  }

  public void setOnItemClickListener(OnItemClickListener l) {
    this.mOnItemClickListener = l;
  }

  @Override
  public int getItemCount() {
    return mDatas != null ? mDatas.size() : 0;
  }

  @SuppressWarnings("unchecked")
  protected final Class<T> getGenericClass() {
    Type genType = getClass().getGenericSuperclass();
    Type[] params = ((ParameterizedType) genType).getActualTypeArguments();

    if (params != null && params.length > 1 && params[1] instanceof Class<?>) {
      return (Class<T>) params[1];
    } else {
      throw new IllegalArgumentException("泛型错误");
    }
  }
}

File: app/src/main/java/remix/myplayer/ui/adapter/BaseAdapter.kt
package remix.myplayer.ui.adapter

import androidx.recyclerview.widget.RecyclerView
import remix.myplayer.misc.interfaces.OnItemClickListener
import remix.myplayer.ui.adapter.holder.BaseViewHolder
import java.lang.reflect.Constructor

/**
 * @ClassName
 * @Description
 * @Author Xiaoborui
 * @Date 2016/10/19 11:31
 */
abstract class BaseAdapter<Data, ViewHolder : BaseViewHolder>(val layoutId: Int) : RecyclerView.Adapter<ViewHolder>() {
  var onItemClickListener: OnItemClickListener? = null
  val dataList: ArrayList<Data> = ArrayList()
  private var constructor: Constructor<*>? = null

  fun setDataList(dataList: List<Data>?) {
    this.dataList.clear()
    if (dataList != null) {
      this.dataList.addAll(dataList)
    }
    notifyDataSetChanged()
  }

  fun getDataList(): List<Data> {
    return dataList
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    convert(holder, getItem(position), position)
  }

  protected open fun getItem(position: Int): Data? {
    return dataList[position]
  }

  protected abstract fun convert(holder: ViewHolder, data: Data?, position: Int)

//  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
//    return try {
//      val itemView = LayoutInflater.from(parent.context).inflate(layoutId, parent, false)
//      constructor?.newInstance(itemView) as ViewHolder
//    } catch (e: Exception) {
//      throw IllegalArgumentException(e.toString())
//    }
//  }

//  fun setOnItemClickListener(l: OnItemClickListener?) {
//    onItemClickListener = l
//  }

  override fun getItemCount(): Int {
    return dataList.size
  }

//  private val genericClass: Class<ViewHolder>
//    get() {
//      val genType = javaClass.genericSuperclass
//      val params = (genType as ParameterizedType).actualTypeArguments
//      return if (params != null && params.size > 1 && params[1] is Class<*>) {
//        params[1] as Class<ViewHolder>
//      } else {
//        throw IllegalArgumentException("泛型错误")
//      }
//    }
//
//  init {
//    try {
//      constructor = genericClass.getDeclaredConstructor(View::class.java)
//      constructor?.isAccessible = true
//    } catch (e: Exception) {
//      throw IllegalArgumentException(e.toString())
//    }
//  }
}